In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
path = r'/Users/jakubkondelka/Documents/CF_Instacart Basket Analysis'

In [3]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [10]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [12]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [14]:
# creates new test data frame 
df_test = pd.DataFrame()

In [15]:
# create a column MIX and fills it with multiple data types
df_test['mix'] = ['a', 'b', 1, True]

In [16]:
# for cycle in Python - example. Runs through all columns and searches for mixed data types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [17]:
df_test['mix'] = df_test['mix'].astype('str')

In [32]:
# display NULL values and SUM them up
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [31]:
#new subset that shows only the missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [21]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [30]:
# removing MISSING values by creating a subset WITHOUT them
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [23]:
df_prods_clean

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49677 rows x 5 columns]

In [29]:
# finding duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [25]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [27]:
df_prods_clean.shape

(49677, 5)

In [28]:
# create subset without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [33]:
df_prods_clean_no_dups.shape

(49672, 5)

In [34]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# TASK 4.5 

In [4]:
# Import libraries
import pandas as pd
import numpy as np
import os
path = r'/Users/jakubkondelka/Documents/CF_Instacart Basket Analysis'

In [47]:
# import orders and cleaned products
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)
df_prods_checked = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index_col = False)
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [7]:
# run describe() on df_prods

df_prods_checked.describe()

Unnamed: 0    product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000  49672.000000   49672.000000  49672.000000
mean   24852.005053  24850.349775     67.762442      11.728942      9.993282
std    14342.265579  14340.705287     38.315784       5.850779    453.615536
min        0.000000      1.000000      1.000000       1.000000      1.000000
25%    12432.750000  12432.750000     35.000000       7.000000      4.100000
50%    24851.500000  24850.500000     69.000000      13.000000      7.100000
75%    37272.250000  37268.250000    100.000000      17.000000     11.100000
max    49692.000000  49688.000000    134.000000      21.000000  99999.000000

Data is consisten, apart from max price being too high, or is just mistaken number. 

In [15]:
# 3 check for mixed data types
# using for loop comand from the exercise 

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

there are no mixed types values

In [17]:
# 5 Run a check for missing values in your df_ords dataframe

df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
evaluations                    0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

I used method that sums up all values that are NULL. days_since_prior_order hasd 206209 missing values 

6 .Address the missing values using an appropriate method.
In a markdown cell, explain why you used your method of choice

In [20]:
# I first create a subset that contains only the missing values 

df_miss = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [21]:
df_miss

Unnamed: 0  order_id  user_id evaluations  order_number  \
0                 0   2539329        1       prior             1   
11               11   2168274        2       prior             1   
26               26   1374495        3       prior             1   
39               39   3343014        4       prior             1   
45               45   2717275        5       prior             1   
...             ...       ...      ...         ...           ...   
3420930     3420930    969311   206205       prior             1   
3420934     3420934   3189322   206206       prior             1   
3421002     3421002   2166133   206207       prior             1   
3421019     3421019   2227043   206208       prior             1   
3421069     3421069   3154581   206209       prior             1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
11                        2                 11                     NaN  
26                        1                 14                     NaN  
39                        6                 11                     NaN  
45                        3                 12                     NaN  
...                     ...                ...                     ...  
3420930                   4                 12                     NaN  
3420934                   3                 18                     NaN  
3421002                   6                 19                     NaN  
3421019                   1                 15                     NaN  
3421069                   3                 11                     NaN  

[206209 rows x 8 columns]

since it is a numeric value, I will replace this with either mean or median. Another way would be to create a column for new customers, because 0 days from prior order can also mean that this was a new customer.

In [24]:
# mean is 1.1 and median is

df_ords['days_since_prior_order'].median()

7.0

I am going to replace with median, because that seems like a value that is more realistic in our example

In [26]:
# replace with median 

df_ords['days_since_prior_order'].fillna(7, inplace=True)

In [29]:
# double check if values were filled in
df_ords.isnull().sum()

Unnamed: 0                0
order_id                  0
user_id                   0
evaluations               0
order_number              0
orders_day_of_week        0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

8. Run a check for duplicates in df_ords

In [33]:
# command for finding duplicates
df_duplicates = df_ords[df_ords.duplicated()]

In [34]:
df_duplicates

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, evaluations, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [32]:
df_ords.shape

(3421083, 8)

there are no duplicates in df_ords

In [35]:
# exporting . df_prods is already exported, so I am going to export df_ords



In [37]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [38]:
df_ords

Unnamed: 0  order_id  user_id evaluations  order_number  \
0                 0   2539329        1       prior             1   
1                 1   2398795        1       prior             2   
2                 2    473747        1       prior             3   
3                 3   2254736        1       prior             4   
4                 4    431534        1       prior             5   
...             ...       ...      ...         ...           ...   
3421078     3421078   2266710   206209       prior            10   
3421079     3421079   1854736   206209       prior            11   
3421080     3421080    626363   206209       prior            12   
3421081     3421081   2977660   206209       prior            13   
3421082     3421082    272231   206209       train            14   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     7.0  
1                         3                  7                    15.0  
2                         3                 12                    21.0  
3                         4                  7                    29.0  
4                         4                 15                    28.0  
...                     ...                ...                     ...  
3421078                   5                 18                    29.0  
3421079                   4                 10                    30.0  
3421080                   1                 12                    18.0  
3421081                   1                 12                     7.0  
3421082                   6                 14                    30.0  

[3421083 rows x 8 columns]

In [42]:
df_ords['order_hour_of_day'].mean()



13.45201534134074

In [43]:
df_ords['order_hour_of_day'].median()


13.0

In [44]:
df_ords['order_hour_of_day'].min()


0

In [45]:
df_ords['order_hour_of_day'].max()

23